[SYCL] Basic diagnostics for the sycl_kernel_entry_point attribute.

tahonermann opened the pull request:
The `sycl_kernel_entry_point` attribute is used to declare a function that defines a pattern for an offload kernel entry point. The attribute requires a single type argument that specifies a class type that meets the requirements for a SYCL kernel name as described in section 5.2, "Naming of kernels", of the SYCL 2020 specification. A unique kernel name type is required for each function declared with the attribute. The attribute may not first appear on a declaration that follows a definition of the function. The function is required to have a `void` return type. The function must not be a non-static member function, be deleted or defaulted, be declared with the `constexpr` or `consteval` specifiers, be declared with the `[[noreturn]]` attribute, be a coroutine, or accept variadic arguments.

Diagnostics are not yet provided for the following:
- Use of a type as a kernel name that does not satisfy the forward declarability requirements specified in section 5.2, "Naming of kernels", of the SYCL 2020 specification.
- Use of a type as a parameter of the attributed function that does not satisfy the kernel parameter requirements specified in section 4.12.4, "Rules for parameter passing to kernels", of the SYCL 2020 specification (each such function parameter constitutes a kernel parameter).
- Use of language features that are not permitted in device functions as specified in section 5.4, "Language restrictions for device functions", of the SYCL 2020 specification.

There are several issues noted by various FIXME comments.
- The diagnostic generated for kernel name conflicts needs additional work to better detail the relevant source locations; such as the location of each declaration as well as the original source of each kernel name.
- A number of the tests illustrate spurious errors being produced due to attributes that appertain to function templates being instantiated too early (during overload resolution as opposed to after an overload is selected).

tahonermann committed:
[SYCL] Basic diagnostics for the sycl_kernel_entry_point attribute.

The `sycl_kernel_entry_point` attribute is used to declare a function that
defines a pattern for an offload kernel entry point. The attribute requires
a single type argument that specifies a class type that meets the requirements
for a SYCL kernel name as described in section 5.2, "Naming of kernels", of
the SYCL 2020 specification. A unique kernel name type is required for each
function declared with the attribute. The attribute may not first appear on a
declaration that follows a definition of the function. The function is
required to have a `void` return type. The function must not be a non-static
member function, be deleted or defaulted, be declared with the `constexpr` or
`consteval` specifiers, be declared with the `[[noreturn]]` attribute, be a
coroutine, or accept variadic arguments.

Diagnostics are not yet provided for the following:
- Use of a type as a kernel name that does not satisfy the forward
  declarability requirements specified in section 5.2, "Naming of kernels",
  of the SYCL 2020 specification.
- Use of a type as a parameter of the attributed function that does not
  satisfy the kernel parameter requirements specified in section 4.12.4,
  "Rules for parameter passing to kernels", of the SYCL 2020 specification
  (each such function parameter constitutes a kernel parameter).
- Use of language features that are not permitted in device functions as
  specified in section 5.4, "Language restrictions for device functions",
  of the SYCL 2020 specification.

There are several issues noted by various FIXME comments.
- The diagnostic generated for kernel name conflicts needs additional work
  to better detail the relevant source locations; such as the location of
  each declaration as well as the original source of each kernel name.
- A number of the tests illustrate spurious errors being produced due to
  attributes that appertain to function templates being instantiated too
  early (during overload resolution as opposed to after an overload is
  selected).

tahonermann committed:
Squash; Fix clang-format complaints.

erichkeane commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -12408,6 +12408,29 @@ def err_sycl_special_type_num_init_method : Error<
+def err_sycl_entry_point_invalid : Error<
+  "'sycl_kernel_entry_point' attribute cannot be applied to a"
+  " %select{non-static member|variadic|deleted|defaulted|constexpr|consteval|"
+           "noreturn|coroutine}0 function">;
```
is `coroutine function` the term of art?  I thought it was just `coroutine`?
```suggestion
           "noreturn|coroutine}0%select{||||||| function}0">;
```

erichkeane replied:
Yeah, that extra select is kind of ugly, I agree.  The alternative is to repeat ` function` on all the other alternatives.

tahonermann replied:
Done. I avoided the additional `%select` by duplicating " function" for the relevant cases. I also changed the wording for the noreturn case from "no return function" to "function declared with the 'noreturn' attribute".

erichkeane commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -12408,6 +12408,29 @@ def err_sycl_special_type_num_init_method : Error<
+  "'sycl_kernel_entry_point' kernel name argument conflicts with a previous"
+  " declaration">;
+def warn_sycl_kernel_name_not_a_class_type : Warning<
+  "%0 is not a valid SYCL kernel name type; a class type is required">,
```
Do we mean 'class type'? As that includes `union`.  Also, I find myself wondering, do the restrictions make sense to exclude scoped-enum?

erichkeane replied:
That would be acceptable, I might consider `class or struct type`, which doesn't get into "what C++ has to say about it" by not using a term of art, and seems similarly descriptive.

tahonermann replied:
Done. I stuck with "non-union class" since the SYCL specification just says "class". We can revisit if/when the SYCL specification is updated to better specify its intent.

erichkeane commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -15978,6 +15988,24 @@ Decl *Sema::ActOnFinishFunctionBody(Decl *dcl, Stmt *Body,
+    } else if (FSI->isCoroutine()) {
+      Diag(FD->getAttr<SYCLKernelEntryPointAttr>()->getLocation(),
+           diag::err_sycl_entry_point_invalid)
+          << /*coroutine*/ 7;
```
When these get this high, an enum typically makes sense instead of comments (which get lost/etc).

tahonermann replied:
Can you point me to an example? I haven't been able to find one. It doesn't look like there is a way to declare an enum with the diagnostic definition. Without the ability to do so, I don't see how adding an intermediate enum somewhere else improves the situation.

erichkeane replied:
Yes, there isn't a convenient way of doing so in the diagnostics engine unfortunately.  I'd love for us to have that!  But typically we just define it elsewhere.  For large lists like this it can make it significantly more readable, and much less likely to not be updated because of a copy/paste error (which happens OFTEN).

See `AttributeArgumentNType` and `AttributeDeclKind` for some examples, but they are kind of sprinkled in quite a few places, Richard had me do some a few times IIRC for function multiversioning.  `Sema::checkTargetVersionAttr` has some more local versions of it.

tahonermann replied:
Thanks for pointing me to those examples. For both `AttributeArgumentNType` and `AttributeDeclKind`, the production of the diagnostic is more complicated. For the first case, the diagnostic is produced in a lambda expression and the `%select` index is passed as an argument. For the second case, the `%select` index is computed by mapping from another enumeration using a switch statement. An enum definitely seems warranted for those cases. Similarly, for `Sema::checkTargetVersionAttr`, there are multiple `%select` indices involved in each diagnostic and the local enumeration does help with readability.

For simple diagnostics, where the `%select` index doesn't require computation and where the same index values are not used in multiple places, I think indirection through an enumeration isn't helpful, so I haven't made a change. If you still think a change is warranted, perhaps we can get a second opinion.

erichkeane replied:
I still disagree strongly.  Any time where there are more than 2-3 options, and are used in more than 1 or 2 places, it vastly improves readability and prevents bugs of printing the wrong thing.  We can have a second opinion from Aaron, but this is something pretty important to me as code-owner.

tahonermann replied:
Thanks, @AaronBallman. I like the suggested tablegen approach and would be happy to use that if someone (other than me) implemented it!

I share your perspective on one-shot enumerations. I'll leave the code as is for now. @erichkeane, as a compromise, I would be happy to split this diagnostic into eight distinct ones to avoid the situation all together. I don't think we gain much by using a single diagnostic in the first place. Let me know if you would prefer that.

AaronBallman replied:
I'd prefer we don't split into separate diagnostics; that increases overhead and maintenance burden for very little gain IMO. We generally prefer using `%select` when possible (it makes it easier to reword diagnostics, fix typos, localize, etc and it avoids adding extra strings to the binary when string merging isn't possible).

erichkeane commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -15978,6 +15988,24 @@ Decl *Sema::ActOnFinishFunctionBody(Decl *dcl, Stmt *Body,
   }
 
+  // Diagnose invalid SYCL kernel entry point function declarations.
+  if (FD && !FD->isInvalidDecl() && !FD->isTemplated() &&
```
Why `isTemplated`? All of these things look like rules you can check even on a dependent function, right? Also, 'isTemplated' just means one of its parents is at template, so in the case of a 'friend' this could actually be defined/declared even without the template being instantiated.

erichkeane replied:
Right, but `isTemplated` doesn't mean `isTemplateDecl`.  So this includes non-templates.

tahonermann replied:
It turned out to be pretty straight forward to enable diagnostics for dependent contexts. No additional changes were needed to diagnose coroutines. For deleted and defaulted functions, I had to duplicate the checks here in `SemaSYCL::CheckSYCLEntryPointFunctionDecl()`; the checks here handle non-dependent cases and the checks in `SemaSYCL::CheckSYCLEntryPointFunctionDecl()` handle dependent cases.

tahonermann replied:
@erichkeane, any further thoughts on this thread? Thanks for your recent comments on other ones!

erichkeane commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -206,3 +208,124 @@ void SemaSYCL::handleKernelEntryPointAttr(Decl *D, const ParsedAttr &AL) {
+    Loc = TT->getDecl()->getLocation();
+  else if (const ObjCInterfaceType *ObjCIT = dyn_cast<ObjCInterfaceType>(T))
+    Loc = ObjCIT->getDecl()->getLocation();
+  return Loc;
```
Should there be an 'assert' in the 'else' here?  It seems like a lot of types here are missing!

tahonermann replied:
No, it is intentional that an invalid source location be returned if there is no source location for the type (e.g., for `int`). I think the above covers all (non-alias) types that can be user declared. I could add comments to better explain the intent for this function; I'll do that.

tahonermann replied:
I renamed the function to `SourceLocationForUserDeclaredType()` and added comments to explain its intended purpose.

erichkeane commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -206,3 +208,124 @@ void SemaSYCL::handleKernelEntryPointAttr(Decl *D, const ParsedAttr &AL) {
+                                QualType KernelName) {
+  assert(!KernelName->isDependentType());
+
+  if (!KernelName->isStructureOrClassType()) {
```
Ah, so enforcing class or structure, and `class` in `[class]` refers to union as well.  So either the SYCL standard is imprecise (and we should probably be more precise with the diagnostic), or this line should really be:

```suggestion
  if (!KernelName->isStructureOrClassType() && !KernelName->isUnionType()) {
```

erichkeane commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -206,3 +208,124 @@ void SemaSYCL::handleKernelEntryPointAttr(Decl *D, const ParsedAttr &AL) {
+  // Ensure that all attributes present on the declaration are consistent
+  // and warn about any redundant ones.
+  const SYCLKernelEntryPointAttr *SKEPAttr = nullptr;
+  for (auto SAI = FD->specific_attr_begin<SYCLKernelEntryPointAttr>();
```
Urgh, I didn't know this `specific_attr_iterator` existed.  I was working at one point to remove a very similar type.

In this case, I'd suggest just using:

`for (auto SAI : llvm::make_filter_range(FD->attrs(), IsaPred<SYCLKernelEntryPointAttr>))`

erichkeane replied:
It has some strange behavior, and is inconsistent with the rest of the iterator types/llvm types that we want to use.  `specific_decl_iterator` is the one I've been working on using.

Also, range-for are much easier to understand, so we tend to have a VAST preference for those unless it is unavoidable.

tahonermann replied:
I swear I had looked for an interface that provided a range of specific/filtered attributes and didn't find one. But it is there! I made a change to use `Decl::specific_attrs()` and a range-based for loop. That is a definite improvement. This still means that `specific_attr_iterator` is used, but as implementation detail. It looks like someone so motivated could change `Decl::specific_attrs()` to return a filter range as you suggest and deprecate/remove the iterator interfaces after transitioning the few uses of those. The changes I made should not require any further modifications to work with a filter range.

erichkeane commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -206,3 +208,124 @@ void SemaSYCL::handleKernelEntryPointAttr(Decl *D, const ParsedAttr &AL) {
+    }
+  }
+
+  if (auto *MD = dyn_cast<CXXMethodDecl>(FD)) {
```
```suggestion
  if (const auto *MD = dyn_cast<CXXMethodDecl>(FD)) {
```

tahonermann replied:
Done.

erichkeane commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -206,3 +208,124 @@ void SemaSYCL::handleKernelEntryPointAttr(Decl *D, const ParsedAttr &AL) {
+    if (!MD->isStatic()) {
+      Diag(SKEPAttr->getLocation(), diag::err_sycl_entry_point_invalid)
+          << /*non-static member function*/ 0;
+    }
```
Also, no curley brackets around all of these 'if' statement bodies as they are single-statement bodies, per coding standard.

tahonermann replied:
Done.

erichkeane commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -206,3 +208,124 @@ void SemaSYCL::handleKernelEntryPointAttr(Decl *D, const ParsedAttr &AL) {
+  if (!SKEPAttr->getKernelName()->isDependentType()) {
+    CheckSYCLKernelName(SemaRef, SKEPAttr->getLocation(),
+                        SKEPAttr->getKernelName());
+  }
```
Single liner here too.

tahonermann replied:
Done. This code received additional changes too.

erichkeane commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -206,3 +208,124 @@ void SemaSYCL::handleKernelEntryPointAttr(Decl *D, const ParsedAttr &AL) {
+    Diag(SKEPAttr->getLocation(), diag::err_sycl_entry_point_return_type);
+  }
+
+  if (!FD->isInvalidDecl() && !FD->isTemplated()) {
```
Once again, do we mean `isTemplated`, or just that this is a template itself?

erichkeane replied:
Again, `isTemplated` doesn't mean the function is a template.  So this would cause problems with things like friend functions in a template context.

tahonermann replied:
`isTemplated()` is correct here and works as expected with hidden friends. `isTemplated()` returns a property of the declaration, not of the declared entity. Consider https://godbolt.org/z/bv1dMs3aq. A non-dependent friend function is defined in a dependent context. The AST dump shows that distinct `FriendDecl` nodes are created for the dependent context and for the specialization of `ct` instantiated to resolve the call from `f()`. I verified in a debugger that `isTemplated()` returns true for the node created for the declaration in the primary template and false for the node created for the instantiation. I added a number of new tests for hidden friends.

tahonermann replied:
@erichkeane, any further thoughts on this thread?

erichkeane commented on clang/lib/Sema/SemaTemplate.cpp:
```diff
@@ -1550,6 +1550,8 @@ NamedDecl *Sema::ActOnNonTypeTemplateParameter(Scope *S, Declarator &D,
     IdResolver.AddDecl(Param);
   }
 
+  ProcessDeclAttributes(S, Param, D);
```
Hmm... this seems like a concerning addition.  I think I want @AaronBallman  to take a look at this, this has implications on appertainment/etc that I don't terribly understand, so I thought this is something we only do in places the standard suggests.

tahonermann replied:
This was needed to ensure diagnostics were issued for a few of the tests. I think this was a simple omission. I appreciate wanting other eyes on it though and we could move this to a separate PR. Note that this checks attributes on the types of non-type template arguments; from the C++ standard perspective, this is like any other type specifier.

erichkeane replied:
Still want Aaron to check on this.

AaronBallman replied:
We definitely need some changes because we drop the attributes on the floor: https://godbolt.org/z/oEr7bneaE

However, this is interesting:
https://eel.is/c++draft/temp#nt:template-parameter
https://eel.is/c++draft/dcl.fct#nt:parameter-declaration

So I would expect us to handle NTTPs the same as other parameter declarations. But... are NTTP's a parameter? If so, our Attr.td and corresponding tablegen might need some updates to distinguish between function parameters and non-function parameters because of things like https://eel.is/c++draft/dcl.attr.depend or https://eel.is/c++draft/dcl.attr.indet because they're specific about function parameters.

So I don't think this code is *wrong*, but I do think we may need to take another look at how we process NTTPs because I would have expected them to use the parameter declaration parsing code, which should automatically handle processing attributes correctly.

tahonermann replied:
Thanks, @AaronBallman.

See also [P3324 (Attributes for namespace aliases, template parameters, and lambda captures)](https://wg21.link/p3324) which would clarify the situation from the perspective of the C++ standard.

What I'll do for now is to move this change to a separate PR and annotate the impacted test (`bad18()` in `clang/test/SemaSYCL/sycl-kernel-entry-point-attr-appertainment.cpp`) with FIXME comments. We can then pursue any additional changes needed via that other PR at our leisure.

tahonermann reviewed (commented):
Responses to Erich's comments. Thanks, Erich!

erichkeane commented on clang/lib/Serialization/ASTReaderDecl.cpp:
```diff
@@ -1134,8 +1134,18 @@ void ASTDeclReader::VisitFunctionDecl(FunctionDecl *FD) {
+    const SYCLKernelInfo *SKI = C.findSYCLKernelInfo(SKEPAttr->getKernelName());
+    if (SKI) {
+      if (!declaresSameEntity(FD, SKI->getKernelEntryPointDecl())) {
+        Reader.Diag(FD->getLocation(), diag::err_sycl_kernel_name_conflict);
```
These tend to be for modules specific stuff, and only as exceptions.  Though upon closer look, this might be for modules-related a diagnostic, but I'll have to take a closer look.

tahonermann replied:
The diagnostic being issued here is specifically for PCH/modules related scenarios. See the `sycl-kernel-entry-point-attr-kernel-name-module.cpp` and `sycl-kernel-entry-point-attr-kernel-name-pch.cpp` tests in `clang/test/SemaSYCL`.

tahonermann committed:
Avoid attempted invalid registration of a SYCL kernel entry point function.

Registration of a function as a SYCL kernel entry point when the kernel name
provided for the `sycl_kernel_entry_point` attribute is invalid could lead to
later errors (e.g., when generating the offload kernel) if the function
declaration is not somehow marked as invalid so that such later attempts can
be skipped. Most attributes, and the `sycl_kernel_entry_point` attribute
specifically, should not affect overload resolution. Marking a function
declaration as invalid with `setInvalidDecl()` causes overload resolution to
skip the declaration, so use of it is not correct for marking function
delcarations with invalid `sycl_kernel_entry_point` attributes. Dropping an
invalid attribute with `dropAttr()` would similarly impact debugging and
diagnostics generation and is therefore not a good option. This change
allows instances of the `sycl_kernel_entry_point` attribute to be marked as
invalid and adds checks to avoid SYCL kernel registration for functions with
an attribute marked as invalid.

Additional tests are provided to validate that the `sycl_kernel_entry_point`
attribute is inherited and that non-dependent invalid uses of the attribute
in dependent contexts are eagerly diagnosed without requiring an instantiation.

tahonermann committed:
Fixes to address the first round of code review comments from Erich.

tahonermann committed:
Squash; Fix clang-format complaints. Again.

tahonermann committed:
Added tests for dependent and non-dependent hidden friend functions.

tahonermann committed:
Fixes to address issues with diagnostics for templated functions, lambdas, and dependent return types.

This change addresses issues with diagnostics generation for various uses of
the sycl_kernel_entry_point attribute:
- Missing diagnostics for deleted or defaulted functions declared in a
  dependent context.
- Missing diagnostics for lambda expressions.
- Spurious diagnostics for functions with a dependent return type; a
  diagnostic was previously issued complaining of a non-void type.
- Missing diagnostics for functions with a deduced return type.

tahonermann committed:
Squash; Fix clang-format complaints. Again.

tahonermann commented:
@erichkeane, thank you for the review comments so far. I think I've addressed or responded to all of them and I look forward to your next round of review. I don't expect you to see this message until January 6th or so, so I hope you are enjoying, or have enjoyed, the holidays!

tahonermann committed:
Update sycl_kernel_entry_point attribute documentation to include the non-deduced void return type requirement.

tahonermann committed:
Remove the fix for attributes in non-type template parameters.

The C++ standard is not clear whether attributes are permitted in the
declarations of non-type template parameters. Previous commits added
code to recognize and instantiate attributes in these contexts, but
concerns were raised that more changes might be needed. This change
removes the previously added support pending clarification from WG21.
See also P3324 (Attributes for namespace aliases, template parameters,
and lambda captures).

erichkeane reviewed (approved):
I think this is fine now.

bader reviewed (approved):
Thanks!

jplehr commented:
Hi @tahonermann do you have an idea on why this is failing?

erichkeane commented:
This must have crossed in the air, we are now rejecting std attributes spelled there.  See https://github.com/llvm/llvm-project/commit/1a73654b3212b623ac21b9deb3aeaadc6906b7e4

jplehr commented:
Can someone more familiar with this push a fix? That would be much appreciated. Thanks!

tahonermann commented:
Thanks for the link to that commit, @erichkeane, I'll push a fix shortly.

tahonermann commented:
A PR to fix the broken test has now been posted at https://github.com/llvm/llvm-project/pull/122375.
